src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
-                    );
-
-                    batch_capabilities.add(&batch.batch_lower);
-
-                    for (pid, offset) in batch.source_upper.iter() {
-                        source_upper_updates.push((*cap.time(), (pid.clone(), *offset)));
+        let work_to_do = tokio::sync::Notify::new();
+        loop {
+            tokio::select! {
+                Some(event) = remap_input.next() => match event {
+                    Event::Data(_cap, data) => {
+                        data.swap(&mut remap_trace_buffer);
+                        remap_updates_stash.append(&mut remap_trace_buffer);
                     }
-                    untimestamped_batches.push_back(batch)
-                }
-            });
-
-            while let Some((_cap, data)) = remap_input.next() {
-                data.swap(&mut remap_trace_buffer);
-                remap_updates_stash.append(&mut remap_trace_buffer);
-            }
-
-            // If the remap frontier advanced it's time to carve out a batch that includes all
-            // updates not beyond the upper
-            let remap_upper = &frontiers[1].frontier();
-            if PartialOrder::less_than(&prev_remap_upper.borrow(), remap_upper) {
-                trace!("reclock({id}) {worker_id}/{worker_count}: remap upper: {remap_upper:?}");
-                let remap_trace_batch = ReclockBatch {
-                    updates: remap_updates_stash
-                        .drain_filter_swapping(|(_, ts, _)| !remap_upper.less_equal(ts))
-                        .collect(),
-                    upper: remap_upper.to_owned(),
-                };
-                timestamper.push_trace_batch(remap_trace_batch);
-                prev_remap_upper = remap_upper.to_owned();
-            }
-
-            // Accumulate updates to bytes_read for Prometheus metrics collection
-            let mut bytes_read = 0;
-            // Accumulate updates to offsets for system table metrics collection
-            let mut metric_updates = HashMap::new();
-
-            trace!(
-                "reclock({id}) {worker_id}/{worker_count}: \
-                untimestamped_batches.len(): {}",
-                untimestamped_batches.len(),
-            );
-
-            while let Some(untimestamped_batch) = untimestamped_batches.front_mut() {
-                // This scope is necessary to convince rustc that `untimestamped_batches` is unused
-                // when we pop from the front at the bottom of this loop.
-                {
-                    // TODO(guswynn&aljoscha): Calculate batch_upper once. :D
-                    let reclock_result =
-                        timestamper.reclock_compat(&mut untimestamped_batch.messages);
-                    let reclocked = match reclock_result {
-                        Ok(reclocked) => reclocked,
-                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
-                            trace!(
```

Comment:
I think in the new code, a trace log line that marks how many batches are not yet ready, tho the `Notify` usage means we will see it way less often...

Reply:
Maybe its worth adding a periodic interval branch to the select that trace logs a bunch of the intermediate state; that kind of information helped us debug prod issues in the past!